File: hospital-hotline-assistant-api/app/services/slip_code.py
"""Slip code derived from a session id.

The patient's printed slip shows a short human-readable code; the nurse at
the destination department types it to pull up the session. Must match the
frontend derivation in ``PatientIdPass.tsx`` (``shortVisitId``) exactly.
"""

from __future__ import annotations

import re

_NON_ALNUM = re.compile(r"[^a-zA-Z0-9]")


def slip_code_for(session_id: str) -> str:
    clean = _NON_ALNUM.sub("", session_id).upper()
    if len(clean) <= 8:
        return clean
    return f"MCH-{clean[:4]}-{clean[-4:]}"

File: hospital-hotline-assistant-web/src/utils/slipCode.ts
/**
 * Slip code shown on the patient's printed slip, derived from the session
 * id. The nurse at the destination department types it to find the session.
 *
 * Must match the backend (`app/services/slip_code.py`) and the patient slip
 * renderer (`PatientIdPass.shortVisitId`) exactly.
 */
export function slipCode(sessionId: string): string {
  const clean = sessionId.replace(/[^a-zA-Z0-9]/g, '').toUpperCase();
  if (clean.length <= 8) return clean;
  return `MCH-${clean.slice(0, 4)}-${clean.slice(-4)}`;
}

/** Normalize a slip code / free text for tolerant matching (drop separators). */
export function slipSearchKey(value: string): string {
  return value.replace(/[^a-zA-Z0-9]/g, '').toUpperCase();
}
